### ISSUE TITLE
Make wrapGAppsHook & co. idempotent

### ISSUE THREAD

**Issue by jtojnar** [NixOS/nixpkgs#57638]
As the [manual](https://nixos.org/nixpkgs/manual/#ssec-setup-hooks) says:

> However, a dependency specified twice will have its setup hook run twice, and that could easily change the build environment (though a well-written setup hook will therefore strive to be idempotent so this is in fact not observable).

This is not maintained by our GNOME setup hooks, especially since we often use `gappsWrapperArgs` directly:

```ShellSession
$ cat $(nix-build -A gvfs)/libexec/gvfsd-fuse | rg GIO_EXTRA_MODULES
export GIO_EXTRA_MODULES='/nix/store/hvdwscbv3w6f5wr7liri89g473v6r6v9-dconf-0.30.1-lib/lib/gio/modules'${GIO_EXTRA_MODULES:+':'}$GIO_EXTRA_MODULES
export GIO_EXTRA_MODULES='/nix/store/hvdwscbv3w6f5wr7liri89g473v6r6v9-dconf-0.30.1-lib/lib/gio/modules'${GIO_EXTRA_MODULES:+':'}$GIO_EXTRA_MODULES
export GIO_EXTRA_MODULES='/nix/store/27yrqjjig6qz3hbkzpkwivmzakas5lw5-gvfs-1.38.1/lib/gio/modules'${GIO_EXTRA_MODULES:+':'}$GIO_EXTRA_MODULES
```

Related: https://discourse.nixos.org/t/declarative-wrappers/1775

cc @hedning @worldofpeace

**Comment by hedning** [comment]
Adding some unique functionality to `addToSearchPath` would be pretty easy. Any good reason not to solve it there?

https://github.com/NixOS/nixpkgs/blob/9034847242eaa034ffa26ef1f75ae25458899c3d/pkgs/stdenv/generic/setup.sh#L187

**Comment by jtojnar** [comment]
We would also need to do the same for `makeWrapper`, or change how the `gappsWrapperArgs`.

Also how would it behave when we want to prioritize certain dirs (own first)

**Comment by hedning** [comment]
Right, might there's places that adds the variables through `gappWrapperArgs` (eg. `find_gio_modules()`), so makeWrapper is perhaps better.

That could also solve prioritization by pruning the prefix and suffix by themselves.

**Comment by qweered** [comment]
triaged, this line is still in manual
